src/ripple/basics/SlabAllocator.h

```diff
@@ -30,67 +30,177 @@
+    // in the slab's allocated block.
+    //
+    // By way of example, on a system with 1 byte long pointers, where the
+    // slab allocator allocates 4 byte long items from a 16 byte long slab
+    // the memory layout is as follows:
+    //
+    //     slot: 0      1      2      3
+    //           [ .... | .... | .... | ....]
+    //
+    // We don't need anything too complicated: the only operations that we
+    // perform are insertion and removal, but only at the head of the list
+    // which means that all we need is very basic CAS.
+    //
+    // Normally, to keep a linked list we need a "node" structure, similar
+    // to this. And, if we're going to have a node, we need some memory in
+    // which to "store" it.
+    //
+    // But remember that we only need a "node" when an item is going to be
+    // on the free list, and something is on the free list because there's
+    // some memory that is otherwise unused.
+    //
+    // So if the size of the unused memory is greater than the size of our
+    // node (which is the size of a pointer) then we can use that directly
+    // and overlay a singly linked list of free objects onto the memory of
+    // those free objects.
+    //
+    // Here's what the memory may look like if slots 1 and 3 are allocated
+    // and slots 0 and 2 are free, with the free list overlaid on top:
+    //
+    //     list_ --------------+
+    //                         |
+    //                         v
+    //     slot: 0      1      2      3
+    //           [ N... | .... | N... | .... ]
+    //           ^ |             |
+    //           | '---> NULL    |
+    //           |               |
+    //           '---------------'
+    //
+    // Notice that although the first empty slot in memory order is slot 0
+    // the free list points to slot 2 as the first available slot and slot
+    // 2 points to slot 0, which points to NULL, as it is it the last free
+    // slot.
+    //
+    // Now imagine that the object in slot 1 is released. The free list is
+    // updated to point to slot 1 as the first available slot. The  memory
+    // now looks like this:
+    //
+    //  list_ ----------+
+    //                  |
+    //                  v
+    //     slot: 0      1      2      3
+    //           [ N... | N... | N... | .... ]
+    //           ^ |      |    ^ |
+    //           | |      |    | |
+    //           | |      '----' |
+    //           | '---> NULL    |
+    //           |               |
+    //           '---------------'
+    //
+    // Note that the order of slots within the free list at any given time
+    // is random, as it depends on the pattern and timing of memory access
+    // of the application.
+    struct alignas(Type) FreeItem
+    {
+        FreeItem* next;
+
+        FreeItem(FreeItem* n = nullptr) noexcept : next(n)
+        {
+        }
+
+        ~FreeItem() noexcept = default;
+    };
+
+    static_assert(sizeof(Type) >= sizeof(FreeItem));
```

Comment:
Should we also assert that alignof(Type) >= alignof(FreeItem)? If I do that it's failing on clang.

Reply:
You know, between the uglyness with the destructors and the alignment issues, I wonder if we're better off without the `FreeItem` class after all.

Reply:
How do you propose to store the `next` pointers without `FreeItem` class?

Reply:
`memcpy` or some function that wraps `memcpy` to make it clearer we're dealing with a linked list. A previous patch did that.

Reply:
Yeah, this was silly of me: on LP64 systems, pointers will end up being aligned on 8 byte boundaries, but `SHAMapItem` is perfectly content with 4 byte alignment.

I _think_ a reasonable solution is to just use an appropriate `alignas()` for `SHAMapItem` (which we can easily check in the slab allocator with a `static_assert`) because the code is a _LOT_ clearer using `FreeItem` instead of the weird/confusing `std::memcpy` syntax.

That said, I'm not opposed to that either.